File: sdk/go/worker.go
package magic

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"sync"
	"time"
)

type HandlerFunc func(input map[string]any) (map[string]any, error)

type capability struct {
	Name           string  `json:"name"`
	Description    string  `json:"description"`
	EstCostPerCall float64 `json:"est_cost_per_call"`
}

type Worker struct {
	name         string
	endpoint     string
	maxWorkers   int
	capabilities map[string]capability
	handlers     map[string]HandlerFunc
	workerID     string
	client       *Client
	sem          chan struct{}
	mu           sync.Mutex
}

func NewWorker(name, endpoint string, maxWorkers int) *Worker {
	return &Worker{
		name:         name,
		endpoint:     endpoint,
		maxWorkers:   maxWorkers,
		capabilities: make(map[string]capability),
		handlers:     make(map[string]HandlerFunc),
		sem:          make(chan struct{}, maxWorkers),
	}
}

func (w *Worker) Capability(name, description string, estCost float64, fn HandlerFunc) {
	w.capabilities[name] = capability{Name: name, Description: description, EstCostPerCall: estCost}
	w.handlers[name] = fn
}

func (w *Worker) HandleTask(taskType string, input map[string]any) (map[string]any, error) {
	fn, ok := w.handlers[taskType]
	if !ok {
		return nil, fmt.Errorf("no handler for %s", taskType)
	}
	return fn(input)
}

func (w *Worker) Register(magicURL, apiKey string) error {
	w.client = NewClient(magicURL, apiKey)
	caps := make([]capability, 0, len(w.capabilities))
	for _, c := range w.capabilities {
		caps = append(caps, c)
	}
	payload := map[string]any{
		"name":         w.name,
		"capabilities": caps,
		"endpoint":     map[string]string{"type": "http", "url": w.endpoint},
		"limits":       map[string]any{"max_concurrent_tasks": w.maxWorkers},
	}
	result, err := w.client.RegisterWorker(payload)
	if err != nil {
		return err
	}
	if id, ok := result["id"].(string); ok {
		w.workerID = id
	}
	slog.Info("registered", "worker_id", w.workerID)
	return nil
}

func (w *Worker) startHeartbeat() {
	go func() {
		for range time.Tick(30 * time.Second) {
			if w.client != nil && w.workerID != "" {
				if err := w.client.Heartbeat(w.workerID); err != nil {
					slog.Warn("heartbeat failed", "err", err)
				}
			}
		}
	}()
}

func (w *Worker) Serve(addr string) error {
	w.startHeartbeat()
	mux := http.NewServeMux()
	mux.HandleFunc("/", w.handleHTTP)
	slog.Info("serving", "name", w.name, "addr", addr)
	return http.ListenAndServe(addr, mux)
}

func (w *Worker) handleHTTP(rw http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(rw, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	var body map[string]any
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		http.Error(rw, "bad json", http.StatusBadRequest)
		return
	}
	msgType, _ := body["type"].(string)
	payload, _ := body["payload"].(map[string]any)
	if msgType != "task.assign" {
		http.Error(rw, "unknown type", http.StatusNotFound)
		return
	}

	taskID, _ := payload["task_id"].(string)
	taskType, _ := payload["task_type"].(string)
	input, _ := payload["input"].(map[string]any)

	// acquire semaphore (non-blocking)
	select {
	case w.sem <- struct{}{}:
	default:
		respond(rw, map[string]any{
			"type":    "task.fail",
			"payload": map[string]any{"task_id": taskID, "error": map[string]string{"code": "overloaded", "message": "at max capacity"}},
		})
		return
	}
	defer func() { <-w.sem }()

	result, err := w.HandleTask(taskType, input)
	if err != nil {
		respond(rw, map[string]any{
			"type":    "task.fail",
			"payload": map[string]any{"task_id": taskID, "error": map[string]string{"code": "handler_error", "message": err.Error()}},
		})
		return
	}
	respond(rw, map[string]any{
		"type":    "task.complete",
		"payload": map[string]any{"task_id": taskID, "output": result, "cost": 0.0},
	})
}

func respond(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

File: sdk/typescript/src/worker.ts
/** MagiC Worker — registers capabilities and handles tasks via HTTP. */

import { createServer, type IncomingMessage, type ServerResponse } from "node:http";

/** Strip trailing slashes without regex (avoids ReDoS). */
function stripTrailingSlashes(s: string): string {
  let end = s.length;
  while (end > 0 && s[end - 1] === "/") end--;
  return s.slice(0, end);
}

interface WorkerOptions {
  name: string;
  endpoint: string;
  port?: number;
}

interface CapabilityEntry {
  name: string;
  description: string;
  handler: (input: Record<string, unknown>) => Promise<unknown>;
}

export class Worker {
  readonly name: string;
  readonly endpoint: string;
  private port: number;
  private capabilities = new Map<string, CapabilityEntry>();
  private workerId?: string;

  constructor(options: WorkerOptions) {
    this.name = options.name;
    this.endpoint = options.endpoint;
    this.port = options.port ?? this.parsePort(options.endpoint);
  }

  private parsePort(endpoint: string): number {
    const match = endpoint.match(/:(\d+)/);
    return match ? parseInt(match[1], 10) : 9000;
  }

  capability(name: string, description: string, handler: (input: Record<string, unknown>) => Promise<unknown>): void {
    this.capabilities.set(name, { name, description, handler });
  }

  async register(serverURL: string): Promise<void> {
    const payload = {
      name: this.name,
      capabilities: Array.from(this.capabilities.values()).map(({ name, description }) => ({ name, description })),
      endpoint: { type: "http", url: this.endpoint },
    };
    const res = await fetch(`${stripTrailingSlashes(serverURL)}/api/v1/workers/register`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      throw new Error(`Registration failed: ${res.status} ${await res.text()}`);
    }
    const data = (await res.json()) as { id?: string };
    this.workerId = data.id;
    console.log(`Registered as ${this.workerId} with ${this.capabilities.size} capabilities`);
  }

  serve(): void {
    const server = createServer(async (req: IncomingMessage, res: ServerResponse) => {
      if (req.method !== "POST") {
        res.writeHead(405).end();
        return;
      }

      let body: string;
      try {
        body = await this.readBody(req);
      } catch {
        res.writeHead(400).end("Bad request");
        return;
      }

      let msg: { type?: string; payload?: Record<string, unknown> };
      try {
        msg = JSON.parse(body);
      } catch {
        res.writeHead(400).end("Invalid JSON");
        return;
      }

      if (msg.type !== "task.assign") {
        res.writeHead(404).end(`Unknown message type: ${msg.type}`);
        return;
      }

      const payload = msg.payload ?? {};
      const taskId = (payload.task_id as string) ?? "unknown";
      const taskType = (payload.task_type as string) ?? "";
      const input = (payload.input as Record<string, unknown>) ?? {};

      let response: unknown;
      const cap = this.capabilities.get(taskType);
      if (!cap) {
        response = {
          type: "task.fail",
          payload: { task_id: taskId, error: { code: "no_handler", message: `No handler for ${taskType}` } },
        };
      } else {
        try {
          const result = await cap.handler(input);
          response = {
            type: "task.complete",
            payload: { task_id: taskId, output: typeof result === "string" ? { result } : result, cost: 0 },
          };
        } catch (e: unknown) {
          console.error("Task handler failed", { taskType, taskId, error: e });
          response = {
            type: "task.fail",
            payload: { task_id: taskId, error: { code: "handler_error", message: "Task handler failed" } },
          };
        }
      }

      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(response));
    });

    server.listen(this.port, "0.0.0.0", () => {
      console.log(`${this.name} serving on 0.0.0.0:${this.port}`);
    });
  }

  private readBody(req: IncomingMessage): Promise<string> {
    return new Promise((resolve, reject) => {
      const chunks: Buffer[] = [];
      req.on("data", (chunk: Buffer) => chunks.push(chunk));
      req.on("end", () => resolve(Buffer.concat(chunks).toString()));
      req.on("error", reject);
    });
  }
}
